ddl/column.go

```diff
@@ -427,13 +427,106 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	originalState := tblInfo.State
+	switch tblInfo.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
```

Comment:
But allowing this differences does not seem to be a big problem. If so there is no need to use any internal state when reorder column, we can change it directly. The only situation which need internal state is convert column data type, because we need backfill all column in such ddl job.